### PR TITLE
Fix support for indefinite timeout.

### PR DESCRIPTION
**Pull request by daku**
Removing commit that I had inadvertently included.

### CODE REVIEW

**Comment by daku** [comment]
@billmalarky , thank you so much for all your help so far. Could I request your attention on this PR when you get a chance.

**Comment by billmalarky** [comment]
Yes, I haven't forgotten about this. Getting a chance to look into it now.

**Comment by billmalarky** [comment]
I went ahead and patched this in my own branch because the default job timeout value must be 25000 instead of 0 like you have it in this PR. The reason jobs default to having a (long) timeout instead of no timeout is because queues started with a lifespan will only pick up jobs with set timeouts. I have a feeling a lot of people who would use react native queue in a background thread (which are limited to 30s execution by iOS and so must have a queue lifespan set) would not see that they need to set a job timeout in order for jobs to get picked up by the queue, and would just assume react native queue is broken. 

So basically, I enforce a job timeout being set unless the developer explicitly chooses for a job _not_ to have a timeout to prevent devs from unknowingly falling into this trap.

https://github.com/billmalarky/react-native-queue/pull/16

**Comment by daku** [comment]
Great thank you. I'll go ahead and pull in the latest version now. Thanks
so much again for the call backs as well.

**Comment by billmalarky** [comment]
I just published the updates to NPM a minute ago, so make sure you didn't get an old version of the package from npm!

Latest is 1.2.1.

https://github.com/billmalarky/react-native-queue/issues/15#issuecomment-382483274

**Comment by billmalarky** [comment]
Also, thank you for catching and reporting this bug!

**Comment by daku** [comment]
Happy to do it and thanks again.

One thing I wonder, if you have time to answer, how do you actually kill a
job that's past it's timeout? Is the actual transaction stopped somehow? I
wasnt able to figure this out from your source code.

**Comment by billmalarky** [comment]
It is not possible to create a "true" timeout in javascript where the execution of the currently executing function itself is killed when a time limit is reached.

So to build this functionality, I spent some time researching the best way to emulate a timeout in javascript, and the best method is to use Promise.race(). Promise.race() accepts promises as arguments, so you can emulate a promise timing out by passing in two promises to Promise.race(). One "real" promise that does the actual job "work" you need done, and a second "timeout" promise, that simply waits the desired timeout length of time before rejecting with a timeout error message. If the job "times out" then promise.race() is fulfilled with the timeout error promise, and if the "real" job completes on time, then promise.race() is fulfilled with the completed job promise.

Note that even if a job "times out" this way, the "real" job promise will continue to execute in the background. This is the nature of the beast with JS, there is no way around this limitation. As a result, this is why react native queue jobs [must be idempotent](https://github.com/billmalarky/react-native-queue#caveats) and react native queue is considered an "at least once" priority queue.

**Comment by daku** [comment]
understood thanks.
